cirq/circuits/expectation_value.py

```diff
@@ -0,0 +1,398 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Expectation value tool for pauli operators given a circuit"""
+
+from typing import Dict, Union
+
+import warnings
+import numpy as np
+from cirq.google import XmonSimulator, XmonMeasurementGate
+from cirq.circuits.circuit import Circuit, Moment
+from cirq.ops import RotXGate, RotYGate, MeasurementGate, Pauli, PauliString, \
+    QubitId
+
+
+def append_measurement_gates(circuit: Circuit,
+                             q_dict: Dict[QubitId,
+                                          Union[int, str]]):
+    """
+    Helper Function to add measurement gates to qubits
+
+    Args:
+        circuit: cirq.Circuit to add measurements to
+        q_dict: dictionary of qubits, keys are integers or strings,
+            values are corresponding qubits
+
+    Returns:
+         circuit with measurements added
+    """
+    meas = [XmonMeasurementGate(key=index).on(
+        qubit) for qubit, index in q_dict.items()]
+
+    circuit.append(meas)
+
+    return circuit
+
+
+def expectation_from_sampling(circuit: Circuit,
+                              operator: Dict[PauliString, float],
+                              n_samples: int):
```

Comment:
This isn't flexible enough for NISQ applications, since we'll often want the number of samples to depend on the PauliString; for instance, see Section IV.A of [arXiv:1801.03524](https://arxiv.org/pdf/1801.03524.pdf). I'm also not convinced that this is the best way for this functionality to be exposed; this is actually being discussed at #982. One big problem with this method as written is that it explicitly uses XmonSimulator, so it's not compatible with other simulators or hardware.

Reply:
It would be convenient to allow for the number of samples to depend on the PauliString. In that case, shouldn't that be an option that one can specify directly when constructing the PauliString? 
While this isn't implemented, wouldn't it still be useful to have the current expectation value functionality exposed? 

Regarding the use of XmonSimulator, do you have any suggestions on an alternate implementation? Of course once hardware or other simulators are released this can also be updated.